das/state.go

```diff
@@ -1,53 +1,223 @@
 package das
 
 import (
-	"sync"
-	"time"
+	"context"
 )
 
-// state collects information about the DASer process. Currently, there are
-// only two sampling routines: the main sampling routine which performs sampling
-// over current network headers, and the `catchUp` routine which performs sampling
-// over past headers from the last sampled checkpoint.
-type state struct {
-	sampleLk sync.RWMutex
-	sample   RoutineState // tracks information related to the main sampling routine
+// coordinatorState represents current state of sampling
+type coordinatorState struct {
+	rangeSize uint64
 
-	catchUpLk sync.RWMutex
-	catchUp   JobInfo // tracks information related to the `catchUp` routine
+	priority   []job                      // list of headers heights that will be sampled with higher priority
+	inProgress map[int]func() workerState // keeps track of running workers
+	failed     map[uint64]int             // stores heights of failed headers with amount of attempt as value
+
+	nextJobID   int
+	next        uint64 // all headers before next were sent to workers
+	networkHead uint64
+
+	catchUpDone   bool          // indicates if all headers are sampled
+	catchUpDoneCh chan struct{} // blocks until all headers are sampled
+}
+
+// initState initiates state for samplingCoordinator
+func initState(samplingRangeSize uint64) coordinatorState {
+	return coordinatorState{
+		rangeSize:     samplingRangeSize,
+		priority:      make([]job, 0),
+		inProgress:    make(map[int]func() workerState),
+		failed:        make(map[uint64]int),
+		nextJobID:     0,
+		next:          1,
+		networkHead:   1,
+		catchUpDone:   false,
+		catchUpDoneCh: make(chan struct{}),
+	}
+}
+
+func (s *coordinatorState) resumeFromCheckpoint(c checkpoint) {
+	s.next = c.SampleFrom
+	s.networkHead = c.NetworkHead
+	// put failed into priority to retry them on restart
+	for h, count := range c.Failed {
+		s.failed[h] = count
+		s.priority = append(s.priority, s.newJob(h, h))
+	}
+}
+
+func (s *coordinatorState) handleResult(res result) {
+	delete(s.inProgress, res.id)
+
+	failedFromWorker := make(map[uint64]bool)
+	for _, h := range res.failed {
+		failedFromWorker[h] = true
+	}
+
+	// update coordinator failed counter
+	for h := res.From; h <= res.To; h++ {
+		if failedFromWorker[h] {
+			// failed in worker, increase failed counter
+			s.failed[h]++
+			continue
+		}
+		delete(s.failed, h)
+	}
```

Comment:
We should loop over `s.failed` instead. In this case, we loop over _each_ sampled header and delete for each, while, in fact, there will be almost none `s.failed` headers. This will save sooo many CPU cycles long run.

Something like this should work
```suggestion
	// check if the worker retried any of the previously failed heights
	for h, c := range s.failed {
		if h < res.From || h > rest.To {
			continue
		}
		
		if failedFromWorker[h] {
			s.failed[h]++
		} else {
			delete(s.failed, h)
		}
		delete(failedFromWorker, h)
	}
	// add newly failed heights
	for h := range failedFromWorker {
		s.failed[h]++
	}
```